content/en/docs/components/pipelines/author-a-pipeline/inputs-outputs-data-passing.md

```diff
@@ -0,0 +1,349 @@
+
+Within a component function body, use input parameters just as you would in a normal Python function.
+
+### Input artifacts
+Input artifacts are defined when you use an `Input[<ArtifactClass>]` annotation. For more information about artifacts, see [Artifacts][artifacts].
+
+At component runtime, input artifacts are copied to the local filesystem by the executing backend. This abstracts away the need for the component author to know where artifacts are stored in remote storage and allows component authors to only interact with the local filesystem when implementing a component that uses an artifact. All artifacts implement a `.path` method, which can be used to access the local path where the artifact file has been copied.
+
+Let's see how this works in practice. In our example pipeline, `augment_dataset` specifies the input `existing_dataset: Input[Dataset]`. In the pipeline definition, we pass the output dataset from `create_dataset` to this parameter. When the `augument_dataset` component runs, the executing backend copies the `output_dataset` artifact file to the container filesystem and passes in an instance of `Dataset` as an argument to `existing_dataset`. The `Dataset` instance has a `.path` handle to its location in the container filesystem, allowing the component to read it:
+
+```python
+with open(existing_dataset.path, 'r') as f:
+    existing_dataset_text = f.read()
+```
+
+## Outputs
+Like inputs, component outputs are also specified by the component function's signature. Depending on the component authoring approach and the type of output (parameter or artifact), outputs may be specified by the function return type annotation (e.g., `-> int`), the type annotation generic `Output[]`, or the type annotation class `OutputPath`. Uses for each are explained in the sections to follow.
+
+For all output types and authoring styles, outputs from a component are persisted to a remote file store, such as [Minio][minio], [Google Cloud Storage][gcs], or [AWS S3][aws-s3], that way they outlast the ephemeral container that creates them and can be picked up for use by a downstream task.
+
+### Output parameters
+Output parameters are declared in different ways depending on the authoring approach.
+
+#### Python components
+For lightweight Python components and containerized Python components, output parameters are declared by the Python component function return type annotation (e.g., `-> int`). Like parameter inputs, return type annotations may be `str`, `int`, `float`, `bool`, `dict` or `list`. 
+
+In our example, `augment_dataset` has a one integer output.
+
+You may also specify multiple output parameters by using these annotations within a `typing.NamedTuple` as follows:
+
+```python
+from kfp import dsl
+from typing import NamedTuple
+
+@dsl.component
+def my_component() -> NamedTuple('Output', [('name', str), ('id', int)]):
+    from typing import NamedTuple
+
+    output = NamedTuple('Output', [('name', str), ('id', int)])
```

Comment:
this should be `collections.namedtuple`, right?

Reply:
Both work, but I prefer `typing.NamedTuple`. Rationale:

All of the examples that I've seem use `typing.NamedTuple` in the annotation then `collections.namedtuple` as the true return type. We request this in our [docs](https://www.kubeflow.org/docs/components/pipelines/sdk/python-function-components/#building-python-function-based-components), too. From my testing, the `typing.NamedTuple` annotation is required, but the return return type can be a `collections` or a `typing` named tuple.

I suspect the examples return a `collections` namedtuple because it's more minimal, not requiring annotations in the `typing.NamedTuple` class factory function, but I think that the confusion of using two different types of named tuples outweighs this "benefit". Furthermore, I'm not convinced that the minimal quality of `collections.namedtuple` is a benefit, as I think using types in both function calls is better for emphasizing the strongly typed and strongly coupled nature of the named tuple class instances. I think these docs are a good opportunity to present a more internally consistent authoring approach. WDYT?